Datatype erasure and traits don't play together

### Dafny version

4.5.0

### Code to produce this issue

```dafny
trait Trait {
  function Value(): string
}

datatype Dt extends Trait = Dt(s: string)
{
  function Value(): string { s }
}

method Main() {
  var d: Dt := Dt("hello");
  var y: Trait := d;
  print d.Value(), " ", y.Value(), "\n";
}
```


### Command to run and resulting output

```code
% dafny run test.dfy --type-system-refresh --general-traits=datatype

Dafny program verifier finished with 2 verified, 0 errors
Errors compiling program into test
source(5726,21): error CS0535: 'Dt' does not implement interface member 'Trait.Value()'
```


### What happened?

Since the datatype has 1 constructor with 1 argument, the compiler optimizes `Dt` to be just that argument. But this optimization should be used only if the datatype does not implement any traits.

### What type of operating system are you experiencing the problem on?

Mac

--- a/Source/DafnyCore/Backends/DatatypeWrapperEraser.cs
+++ b/Source/DafnyCore/Backends/DatatypeWrapperEraser.cs
@@ -136,9 +136,10 @@ namespace Microsoft.Dafny.Compilers {
     ///   2 -- be an inductive datatype (not a "codatatype"), and
     ///   3 -- have exactly one non-ghost constructor, and
     ///   4 -- that constructor must have exactly one non-ghost destructor parameter (say, "d" of type "D"), and
-    ///   5 -- have no fields declared as members, and
-    ///   6 -- the compiled parts of type "D" must not include the datatype itself, and
-    ///   7 -- not be declared with {:extern} (since extern code may rely on it being there).
+    ///   5 -- have no non-ghost fields declared as members, and
+    ///   6 -- have no parent traits, and
+    ///   7 -- the compiled parts of type "D" must not include the datatype itself, and
+    ///   8 -- not be declared with {:extern} (since extern code may rely on it being there).
     ///
     /// If the conditions above apply, then the method returns true and sets the out-parameter to the core DatatypeDestructor "d".
     /// From this return, the compiler (that is, the caller) will arrange to compile type "dt" as type "D".
@@ -161,12 +162,13 @@ namespace Microsoft.Dafny.Compilers {
     }
 
     /// <summary>
-    /// Check for conditions 2, 3, 4, 5, and 7 (but not 0, 1, and 6) mentioned in the description of IsErasableDatatypeWrapper.
+    /// Check for conditions 2, 3, 4, 5, 6, and 8 (but not 0, 1, and 7) mentioned in the description of IsErasableDatatypeWrapper.
     /// </summary>
     private static bool FindUnwrappedCandidate(DafnyOptions options, DatatypeDecl datatypeDecl, out DatatypeDestructor coreDtor) {
       if (datatypeDecl is IndDatatypeDecl &&
           !datatypeDecl.IsExtern(options, out _, out _) &&
-          !datatypeDecl.Members.Any(member => member is Field)) {
+          !datatypeDecl.Members.Any(member => member is Field { IsGhost: false }) &&
+          datatypeDecl.ParentTraits.Count == 0) {
         var nonGhostConstructors = datatypeDecl.Ctors.Where(ctor => !ctor.IsGhost).ToList();
         if (nonGhostConstructors.Count == 1) {
           // there is exactly one non-ghost constructor

--- a/Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/ErasableTypeWrappers.dfy
+++ b/Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/ErasableTypeWrappers.dfy
@@ -1,4 +1,4 @@
-// RUN: %testDafnyForEachCompiler --refresh-exit-code=0 "%s" -- --relax-definite-assignment --spill-translation --unicode-char:false
+// RUN: %testDafnyForEachCompiler --refresh-exit-code=0 "%s" -- --relax-definite-assignment --spill-translation --unicode-char:false --type-system-refresh --general-traits=datatype
 
 datatype SingletonRecord = SingletonRecord(u: int)
 datatype GhostOrNot = ghost Ghost(a: int, b: int) | Compiled(x: int)
@@ -18,6 +18,7 @@ method Main() {
   TestMembers();
   OptimizationChecks.Test();
   PrintRegressionTests.Test();
+  ConditionsThatDisableTheOptimization.Test();
 }
 
 method TestTargetTypesAndConstructors() {
@@ -410,5 +411,35 @@ module PrintRegressionTests {
   method PrintOne<X(0)>() {
     var x: X := *;
     print x, "\n";
+  }
+}
+
+// --------------------------------------------------------------------------------
+
+module ConditionsThatDisableTheOptimization {
+  trait Trait {
+    function F(): int
+  }
+  datatype WithTrait extends Trait = WithTrait(u: int)
+  {
+    function F(): int { 5 }
+  }
+
+  datatype WithCompiledField = WithCompiledField(u: int)
+  {
+    const n: int := 100
+  }
+
+  datatype WithGhostField = WithGhostField(u: int)
+  {
+    ghost const n: int := 100
+  }
+
+  method Test() {
+    var t0 := WithTrait(40);
+    var t1 := WithCompiledField(41);
+    var t2 := WithGhostField(42);
+    print t0, " ", t0.F(), " ", (t0 as Trait).F(), "\n"; // WithTrait(40) 5 5
+    print t1, " ", t1.n, " ", t2, "\n"; // WithCompiledField(41) 100 42
   }
 }

--- a/Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/ErasableTypeWrappers.dfy.expect
+++ b/Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/ErasableTypeWrappers.dfy.expect
@@ -18,3 +18,5 @@ HasConst.MakeC(4) (4, 4)
 OptimizationChecks.Ints.Ints(5, 7) OptimizationChecks.Ints.Ints(5, 7) OptimizationChecks.Ints.AnotherIntsWrapper(OptimizationChecks.Ints.Ints(5, 7))
 D D
 5 5
+ConditionsThatDisableTheOptimization.WithTrait.WithTrait(40) 5 5
+ConditionsThatDisableTheOptimization.WithCompiledField.WithCompiledField(41) 100 42

--- a/docs/dev/news/5234.fix
+++ b/docs/dev/news/5234.fix
@@ -1,0 +1,2 @@
+fix: Disable the "erase datatype wrappers" optimization if the datatype implements any traits.
+feat: Allow the "erase datatype wrappers" optimization if the only fields in the datatype are ghost fields.

